README.md
# Extend String

[![Custom badge](https://img.shields.io/endpoint?url=https%3A%2F%2Fdeno-visualizer.danopia.net%2Fshields%2Flatest-version%2Fx%2Festring%2Fmod.ts)](https://doc.deno.land/https/deno.land/x/estring/mod.ts)
![Deno test](https://github.com/maldan/denolib-extend-string/workflows/Deno/badge.svg)
[![License](https://img.shields.io/github/license/maldan/denolib-extend-string)](https://github.com/maldan/denolib-bytearray/blob/master/LICENSE)
[![hits](https://hits.deltapapa.io/github/maldan/denolib-extend-string.svg)](https://hits.deltapapa.io)
<a href="https://github.com/maldan/denolib-extend-string/pulse" alt="Activity">
<img src="https://img.shields.io/github/commit-activity/m/maldan/denolib-extend-string" />
</a>
![GitHub Repo stars](https://img.shields.io/github/stars/maldan/denolib-extend-string)

Extend standard js string with this module.

## Full documentation

Go here - https://doc.deno.land/https/deno.land/x/estring/mod.ts

## How to import

```ts
import { EString } from "https://deno.land/x/estring@2.0.0/mod.ts";
```

## Function list

### Transform

-   caseStyle

### Validate

-   email

## Example

```ts
EString("hello").transform.capitalize(); // Hello
EString("hello").validate.email(); // false
EString("dd@dd.ru").validate.email(); // true
```

### Have a question/suggestion/bug report?

Contact me: https://t.me/maldan


CHANGELOG.md
## [2.0.0] - 2021-02-26

### Changed

-   Everything

### Added

-   `validate.email`
-   `transofrm.capitalize`
-   `transofrm.caseStyle`
-   `transofrm.timeToInt`


test/base.test.ts
import { assertEquals } from "https://deno.land/std@0.85.0/testing/asserts.ts";
import { StringExtend } from "../src/StringExtend.ts";

Deno.test("capitalize", () => {
    assertEquals(new StringExtend("hello").capitalize, "Hello");
});

Deno.test("digits", () => {
    assertEquals(new StringExtend("x123dac").digits, 123);
});

Deno.test("validate email", () => {
    assertEquals(new StringExtend("dd@dd.ru").isValidEmail, true);
    assertEquals(new StringExtend("ddxdd.ru").isValidEmail, false);
});

Deno.test("replace at", () => {
    assertEquals(new StringExtend("1234").replaceAt(1, "5"), "1534");
});

Deno.test("snake to camel", () => {
    assertEquals(new StringExtend("open_file").snakeToCamel, "openFile");
});

Deno.test("kebab to camel", () => {
    assertEquals(new StringExtend("open-file").kebabToCamel, "openFile");
});

Deno.test("dot to camel", () => {
    assertEquals(new StringExtend("file.open.error").dotToCamel, "fileOpenError");
});

Deno.test("camel to kebab", () => {
    assertEquals(new StringExtend("openFile").camelToKebab, "open-file");
    assertEquals(new StringExtend("open").camelToKebab, "open");
});

Deno.test("camel to dot", () => {
    assertEquals(new StringExtend("openFile").camelToDot, "open.file");
    assertEquals(new StringExtend("open").camelToDot, "open");
});


test/validate.test.ts
import { assertEquals } from "https://deno.land/std@0.85.0/testing/asserts.ts";
import { EString, StringExtend } from "../src/StringExtend.ts";

Deno.test("validate email", () => {
    const correct = ["a@a.ru", "text@com.ua.ru", "hi@hi.com"];
    const inCorrect = ["sdfsdf", "", "23234", "dd@", "@dd.ru"];

    correct.forEach((x) => {
        assertEquals(EString(x).validate.email(), true);
    });
    inCorrect.forEach((x) => {
        assertEquals(EString(x).validate.email(), false);
    });
});

/*Deno.test("capitalize", () => {
    assertEquals(new StringExtend("hello").capitalize, "Hello");
});

Deno.test("digits", () => {
    assertEquals(new StringExtend("x123dac").digits, 123);
});

Deno.test("validate email", () => {
    assertEquals(new StringExtend("dd@dd.ru").isValidEmail, true);
    assertEquals(new StringExtend("ddxdd.ru").isValidEmail, false);
});

Deno.test("replace at", () => {
    assertEquals(new StringExtend("1234").replaceAt(1, "5"), "1534");
});

Deno.test("snake to camel", () => {
    assertEquals(new StringExtend("open_file").snakeToCamel, "openFile");
});

Deno.test("kebab to camel", () => {
    assertEquals(new StringExtend("open-file").kebabToCamel, "openFile");
});

Deno.test("dot to camel", () => {
    assertEquals(new StringExtend("file.open.error").dotToCamel, "fileOpenError");
});

Deno.test("camel to kebab", () => {
    assertEquals(new StringExtend("openFile").camelToKebab, "open-file");
    assertEquals(new StringExtend("open").camelToKebab, "open");
});

Deno.test("camel to dot", () => {
    assertEquals(new StringExtend("openFile").camelToDot, "open.file");
    assertEquals(new StringExtend("open").camelToDot, "open");
});*/


test/transoform.test.ts
import { assertEquals } from "https://deno.land/std@0.85.0/testing/asserts.ts";
import { EString, StringExtend } from "../src/StringExtend.ts";

Deno.test("transofrm capitalize", () => {
    assertEquals(EString("a").transofrm.capitalize(), "A");
    assertEquals(EString("").transofrm.capitalize(), "");
    assertEquals(EString("hi").transofrm.capitalize(), "Hi");
    assertEquals(EString("HAH").transofrm.capitalize(), "HAH");

    console.log(EString("HeLo").transofrm.caseStyle("camel", "kebab"));
});

Deno.test("transofrm case style camel -> any", () => {
    assertEquals(EString("helloWorld").transofrm.caseStyle("camel", "kebab"), "hello-world");
    assertEquals(EString("helloWorld").transofrm.caseStyle("camel", "dot"), "hello.world");
});

Deno.test("transofrm time to int", () => {
    assertEquals(EString("00:30").transofrm.timeToInt("mm:ss"), 30);
    assertEquals(EString("01:30").transofrm.timeToInt("mm:ss"), 90);

    assertEquals(EString("00:01").transofrm.timeToInt("hh:mm"), 60);
    assertEquals(EString("00:02").transofrm.timeToInt("hh:mm"), 120);
    assertEquals(EString("01:00").transofrm.timeToInt("hh:mm"), 60 * 60);
    assertEquals(EString("01:01").transofrm.timeToInt("hh:mm"), 60 * 60 + 60);

    assertEquals(EString("00:00:00").transofrm.timeToInt("hh:mm:ss"), 0);
    assertEquals(EString("00:00:10").transofrm.timeToInt("hh:mm:ss"), 10);
    assertEquals(EString("01:00:00").transofrm.timeToInt("hh:mm:ss"), 60 * 60);
});

/*Deno.test("capitalize", () => {
    assertEquals(new StringExtend("hello").capitalize, "Hello");
});

Deno.test("digits", () => {
    assertEquals(new StringExtend("x123dac").digits, 123);
});

Deno.test("validate email", () => {
    assertEquals(new StringExtend("dd@dd.ru").isValidEmail, true);
    assertEquals(new StringExtend("ddxdd.ru").isValidEmail, false);
});

Deno.test("replace at", () => {
    assertEquals(new StringExtend("1234").replaceAt(1, "5"), "1534");
});

Deno.test("snake to camel", () => {
    assertEquals(new StringExtend("open_file").snakeToCamel, "openFile");
});

Deno.test("kebab to camel", () => {
    assertEquals(new StringExtend("open-file").kebabToCamel, "openFile");
});

Deno.test("dot to camel", () => {
    assertEquals(new StringExtend("file.open.error").dotToCamel, "fileOpenError");
});

Deno.test("camel to kebab", () => {
    assertEquals(new StringExtend("openFile").camelToKebab, "open-file");
    assertEquals(new StringExtend("open").camelToKebab, "open");
});

Deno.test("camel to dot", () => {
    assertEquals(new StringExtend("openFile").camelToDot, "open.file");
    assertEquals(new StringExtend("open").camelToDot, "open");
});*/


src/Transform.ts
import { StringExtend } from "./StringExtend.ts";

type CaseStyleFrom = "camel" | "kebab" | "pascal" | "snake" | "dot";
type CaseStyleTo = "camel" | "kebab" | "flat" | "upper flat" | "pascal" | "snake" | "dot";

export class Transform {
    readonly extend: StringExtend;

    constructor(extend: StringExtend) {
        this.extend = extend;
    }

    capitalize(): string {
        if (!this.extend.string[0]) {
            return "";
        }
        return this.extend.string[0].toUpperCase() + this.extend.string.slice(1);
    }

    timeToInt(mask: "hh:mm" | "hh:mm:ss" | "mm:ss" = "hh:mm") {
        const s = this.extend.string;
        if (mask === "hh:mm") {
            const hms = s + ":00";
            const a = hms.split(":");
            return ~~a[0] * 60 * 60 + ~~a[1] * 60 + ~~a[2];
        }
        if (mask === "hh:mm:ss") {
            const a = s.split(":");
            return ~~a[0] * 60 * 60 + ~~a[1] * 60 + ~~a[2];
        }
        if (mask === "mm:ss") {
            const a = s.split(":");
            return ~~a[0] * 60 + ~~a[1];
        }
        throw new Error(`Unsupported "${mask}" mask!`);
    }

    caseStyle(from: CaseStyleFrom, to: CaseStyleTo): string {
        const s = this.extend.string;

        switch (from) {
            case "camel":
                switch (to) {
                    case "kebab":
                        return s.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase();
                    case "dot":
                        return s.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1.$2").toLowerCase();
                    case "snake":
                        return s.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1_$2").toLowerCase();
                    case "flat":
                        return s.toLowerCase();
                    case "upper flat":
                        return s.toUpperCase();
                    default:
                        throw new Error(`Unknown case "${to}" style!`);
                }
            default:
                throw new Error(`Unknown case "${from}" style!`);
        }
    }
}


src/StringExtend.ts
import { Validate } from "./validate.ts";
import { Transform } from "./transform.ts";

export class StringExtend {
    readonly string: string;
    readonly validate = new Validate(this);
    readonly transofrm = new Transform(this);

    constructor(string: string) {
        this.string = string;
    }

    /*get digits(): number {
        return Number(this._str.replace(/\D/g, ""));
    }

    replaceAt(index: number, replacement: string): string {
        return (
            this._str.substr(0, index) + replacement + this._str.substr(index + replacement.length)
        );
    }

    get capitalize(): string {
        return this._str[0].toUpperCase() + this._str.slice(1);
    }*/

    /*get snakeToCamel(): string {
        return this._str.replace(/(_\w)/g, (x) => x[1].toUpperCase());
    }

    get kebabToCamel(): string {
        return this._str.replace(/(-\w)/g, (x) => x[1].toUpperCase());
    }

    get dotToCamel(): string {
        return this._str.replace(/(\.\w)/g, (x) => x[1].toUpperCase());
    }

    get camelToKebab(): string {
        return this._str.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase();
    }

    get camelToDot(): string {
        return this._str.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1.$2").toLowerCase();
    }*/

    /*get isValidEmail(): boolean {
        const re = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
        return re.test(this._str.toLowerCase());
    }

    get htmlSpecialChars(): string {
        return this._str
            .replace(/&/g, "&amp;")
            .replace(/"/g, "&quot;")
            .replace(/'/g, "&#039;")
            .replace(/</g, "&lt;")
            .replace(/>/g, "&gt;");
    }*/
}

export function EString(str: string): StringExtend {
    return new StringExtend(str);
}


global.ts
declare global {
  interface String {
    /**
         * Convert string with date to date object
         */
    toDate(): Date;

    /**
         * Extract from string all digits
         */
    digits(): number;

    /**
         * Replace char at specific position
         * @param index
         * @param replacement
         */
    replaceAt(index: number, replacement: string): string;

    /**
         * Convert dmy date to ymd
         * @constructor
         */
    DMYToYMD(): string;

    /**
         * Convert ymd to dmy
         * @constructor
         */
    YMDToDMY(): string;

    /**
         * Convert hours:minutes to int. For example "00:59" -> 59 and "01:01" -> 61.
         * @constructor
         */
    HMToInt(): number;

    /**
         * Convert time to human readable time, for example 22:30 -> 22 h 30 min
         */
    humanTime(): string;

    /**
         * Get string between two string. For example "My name is <John> Watson".between("<", ">") will be "John".
         * @param first
         * @param last
         */
    between(first: string, last: string): string;

    /**
         * Get amount of specific char in string. For example "hello".count("l") will be 2.
         * @param char
         */
    count(char: string): number;

    /**
         * Convert snake to camel. Example sex_rock to sexRock
         */
    snakeToCamel(): string;

    /**
         * Convert kebab to camel. Example sex-rock to sexRock
         */
    kebabToCamel(): string;

    /**
         * Convert camel to kebab. Example sexRock to sex-rock
         */
    camelToKebab(): string;

    /**
         * Convert dot to camel. Example sex.rock to sexRock
         */
    dotToCamel(): string;

    capitalize(): string;

    /**
         * Calculate max match length. For example "hello".maxCharsMatch("hell") will be 4.
         * Because "hello" contains a word "hell". If we call "hello".maxCharsMatch("123") we will
         * get 0 because there is no "123" in "hello" word.
         * @param str
         */
    maxCharsMatch(str: string): number;

    htmlSpecialChars(): string;

    dataUrlToBuffer(): string;

    validateEmail(): boolean;

    //leftPad(len: number, ch?: string): string;
  }
}

String.prototype.toDate = function (): Date {
  // eslint-disable-next-line @typescript-eslint/no-this-alias
  let date = this;
  let d;
  if (date.indexOf(" ") !== -1) {
    date = date.replace(" ", "T");
  }
  if (date.indexOf("T") !== -1) {
    d = new Date(date);
  }

  if (isNaN(d.getTime())) {
    return null;
  }

  /*let t = date.split('-');
    if (t.length === 1) t = date.split('.');

    if (t[2] * 1 > 1000)
        d = new Date(Date.UTC(t[2] * 1, t[1] * 1 - 1, t[0] * 1));
    else
        d = new Date(Date.UTC(t[0] * 1, t[1] * 1 - 1, t[2] * 1));

    d.setHours(0, 0, 0, 0);*/
  return d;
};

String.prototype.digits = function () {
  return this.replace(/\D/g, "") * 1;
};

String.prototype.replaceAt = function (index: number, replacement: string) {
  return this.substr(0, index) + replacement +
    this.substr(index + replacement.length);
};

String.prototype.DMYToYMD = function (): string {
  const date = this.split(" ")[0];
  const time = this.split(" ").length > 1 ? (" " + this.split(" ")[1]) : "";
  const tuple = date.split(".");
  return tuple[2] + "-" + tuple[1] + "-" + tuple[0] + time;
};

String.prototype.YMDToDMY = function (): string {
  const date = this.split(" ")[0];
  const time = this.split(" ").length > 1 ? (" " + this.split(" ")[1]) : "";
  const tuple = date.split("-");
  return tuple[2] + "." + tuple[1] + "." + tuple[0] + time;
};

String.prototype.HMToInt = function () {
  let hms = this + ":00";
  let a = hms.split(":");
  // if (a.length === 1) return 0;
  return (~~a[0]) * 60 * 60 + (~~a[1]) * 60 + (~~a[2]);
};

String.prototype.humanTime = function () {
  let tuple = this.split(":");
  tuple[0] = ~~tuple[0];
  tuple[1] = ~~tuple[1];
  if (tuple.length === 3) tuple[2] = ~~tuple[2];

  if (tuple.length === 3) {
    return `${tuple[0]} h ${tuple[1]} min ${tuple[2]} sec`;
  }
  if (tuple.length === 2) return `${tuple[0]} h ${tuple[1]} min`;
};

String.prototype.between = function (first: string, last: string) {
  let f1 = this.indexOf(first), f2 = this.indexOf(last);
  if (f1 === -1 || f2 === -1) return this;
  if (first === last) {
    f2 = this.indexOf(last, 1);
  }
  return this.substr(f1 + first.length, f2 - f1 - last.length);
};

String.prototype.count = function (char: string) {
  let out = 0;
  for (let i = 0; i < this.length; i++) {
    if (this[i] === char) out += 1;
  }
  return out;
};

String.prototype.snakeToCamel = function () {
  return this.replace(/(_\w)/g, (x) => x[1].toUpperCase());
};

String.prototype.kebabToCamel = function () {
  return this.replace(/(-\w)/g, (x) => x[1].toUpperCase());
};

String.prototype.camelToKebab = function () {
  return this.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase();
};

String.prototype.dotToCamel = function () {
  return this.replace(/(\.\w)/g, (x) => x[1].toUpperCase());
};

String.prototype.maxCharsMatch = function (str: string): number {
  for (let i = 0; i < str.length; i++) {
    let part = str.slice(0, -i) || str;
    if (this.match(part)) return part.length;
  }
  return 0;
};

String.prototype.htmlSpecialChars = function (): string {
  return this.replace(/&/g, "&amp;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#039;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
};

String.prototype.validateEmail = function (): boolean {
  const re =
    /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return re.test(this.toLowerCase());
};

String.prototype.capitalize = function (): string {
  return this[0].toUpperCase() + this.slice(1);
};

/*String.prototype.leftPad = function (len: number, ch: string = ' '): string {
    let str = String(this);
    var i = -1;

    len = len - str.length;
    while (++i < len) {
        str = ch + str;
    }
    return str;
}*/

export default class StringExtender {
}
